File: C++/myLeetCode/Binary_Tree/Find the attributes of Binary tree/572. Subtree of Another Tree.cpp
// 572. Subtree of Another Tree
// https://leetcode.com/problems/subtree-of-another-tree/
// g++ 572.\ Subtree\ of\ Another\ Tree.cpp -std=c++17
#include "../../printUtils.hpp"

template <typename T>
bool isSameTree(TreeNode<T>* p, TreeNode<T>* q) {
    if (p == nullptr && q == nullptr) {
        return true;
    }
    if (p == nullptr || q == nullptr) {
        return false;
    }
    if (p->val != q->val) {
        return false;
    }
    return isSameTree(p->left, q->left) && isSameTree(p->right, q->right);
}

template <typename T>
bool isSubtree(TreeNode<T>* s, TreeNode<T>* t) {
    if (s == nullptr) {
        return false;
    }
    if (isSameTree(s, t)) {
        return true;
    }
    return isSubtree(s->left, t) || isSubtree(s->right, t);
}

int main() {
    TreeNode<int>* s = buildTree({3, 4, 5, 1, 2}, -1);
    TreeNode<int>* t = buildTree({4, 1, 2}, -1);

    std::cout << std::boolalpha << isSubtree(s, t) << std::endl;

    return 0;
}

File: C#/myLeetCode/Binary_Tree/Find the attributes of Binary tree/572. Subtree of Another Tree.cs
// 572. Subtree of Another Tree
// https://leetcode.com/problems/subtree-of-another-tree/

using System;

public class TreeNode {
    public int val;
    public TreeNode? left;
    public TreeNode? right;
    public TreeNode(int x) { val = x; }
}

public class Solution {
    public bool IsSubtree(TreeNode? s, TreeNode? t) {
        if (s == null) return false;
        if (IsSameTree(s, t)) return true;
        return IsSubtree(s.left, t) || IsSubtree(s.right, t);
    }

    private bool IsSameTree(TreeNode? s, TreeNode? t) {
        if (s == null && t == null) return true;
        if (s == null || t == null) return false;
        return s.val == t.val && IsSameTree(s.left, t.left) && IsSameTree(s.right, t.right);
    }
}

public class Program {
    public static void Main() {
        Solution solution = new Solution();

        TreeNode tree1 = new TreeNode(3) {
            left = new TreeNode(4) {
                left = new TreeNode(1),
                right = new TreeNode(2)
            },
            right = new TreeNode(5)
        };

        TreeNode tree2 = new TreeNode(4) {
            left = new TreeNode(1),
            right = new TreeNode(2)
        };

        bool isSubtree = solution.IsSubtree(tree1, tree2);

        // 打印结果
        Console.WriteLine($"Is tree2 a subtree of tree1? {isSubtree}");
    }
}
